### ISSUE TITLE
Consistent delay for recieving keypresses after a call to swap_buffer.

### ISSUE THREAD

**Issue by BigfootN** [rust-windowing/glutin#1054]
Hello,

I noticed that, with `alacritty`, based on `glutin`, it takes some time to respond when an (for example) arrow is pressed for a long time. The cursor (for example in `nvim`, even with `nvim -u NONE`) takes some time to stop, it continues for some time.

The reason I am opening an issue here is because I cloned the `glutin` repository and ran 
`cargo run --example window`. I seemed to notice that the output from the terminal from which I ran the command seemed to be going on for a small amount of time despite the fact that all keys where released.

Some system information:

OS: `archlinux`
kernel: `4.17.11-arch1`

I hope that I was clear enough.

Thank you very much in advance.

**Comment by francesca64** [comment]
If this also happens with the [winit](https://github.com/tomaka/winit) examples, then this should be reported to winit instead.

**Comment by BigfootN** [comment]
Hello,

Sorry for the delay of the response, thank you for the suggestion.
So, I tried the examples and I didn't notice anything.

Thank you very much in advance.

**Comment by felixrabe** [comment]
@BigfootN To clarify: So you noticed the delayed response with glutin, but not with winit?

**Comment by BigfootN** [comment]
Sorry for the late response.

So yes, it's what I concluded from the examples.

**Comment by chrisduerr** [comment]
This also seems to reproduce on macOS:
https://github.com/jwilm/alacritty/issues/2514#issuecomment-499538851

**Comment by ZeGentzy** [comment]
@chrisduerr Can I post our irc log from June 11th, for those interested?

**Comment by chrisduerr** [comment]
This is the full log of the discussion:

<details>

```
2019-06-11 18:10:07     UndeadLeech     Does the swap_buffers call on a split context interfere with the event loop with the event loop 2 rework?
2019-06-11 18:10:40     UndeadLeech     Before the change where gentz separated the context from the window+surface.
2019-06-12 00:27:17     gentz   I presume so. It should. If it doesn't, I'd call it a bug.
2019-06-12 00:29:15     gentz   Or more precisely, it should interfere no more or less then it would normaly of had if the two had not been split.
2019-06-12 00:29:19     gentz   UndeadLeech
2019-06-12 00:31:11     UndeadLeech     I'm getting some issues with inconsistent ReceivedCharacter events, I've actually just got a pretty minimal repro case, I just have to sanity check it since I'm a bit slow today.
2019-06-12 00:33:59     UndeadLeech     Okay I'm not sure if there's anything super stupid here, but this is approximately the troublesome code gentz: https://paste.rs/gd2.rs
2019-06-12 00:34:30     UndeadLeech     If I set my key repeat rate to 20 using `xset r rate 200 20`, everything's fine and pretty much every time it prints the time since the last event was 50ms.
2019-06-12 00:35:44     UndeadLeech     However if I set the repeat rate to 200 using `xset r rate 200 200`, I suddenly get loads of fluctuations where it's like sometimes < 5 micro seconds and then other times over 15 ms.
2019-06-12 00:36:34     gentz   Can you reproduces this with winit only?
2019-06-12 00:36:40     gentz   s/s//
2019-06-12 00:39:15     UndeadLeech     I had trouble reconstructing this examlpe with winit.
2019-06-12 00:39:37     UndeadLeech     The whole context stuff is from glutin, right?
2019-06-12 00:39:56     UndeadLeech     I'm not able to reproduce this with the normal examples of winit evlp2/master.
2019-06-12 00:40:01     UndeadLeech     Or the normal examples of glutin.
2019-06-12 00:40:04     gentz   Yeah. Just make the window with the window builder
2019-06-12 00:40:34     UndeadLeech     This only happens when I setup this double-thread system where one thread notifies the other about redraw requests.
2019-06-12 00:40:48     UndeadLeech     It doesn't matter which thread swaps buffers though.
2019-06-12 00:41:33     UndeadLeech     Originally I thought this was just due to the time taken by swap_buffers, but based on my crude benchmarks it doesn't take long enough to justify the 'lags' I'm getting.
2019-06-12 00:42:47     gentz   How does this do:
2019-06-12 00:42:50      *      gentz sent a long message:  < https://matrix.org/_matrix/media/v1/download/matrix.org/awllSXJhlXXNvKnCQjzawKAZ >
2019-06-12 00:42:57     gentz   (I haven't tested if it compiles)
2019-06-12 00:43:52     UndeadLeech     No problems.
2019-06-12 00:44:13     gentz   Hmmm
2019-06-12 00:50:22      *      gentz sent a long message:  < https://matrix.org/_matrix/media/v1/download/matrix.org/xahigaTtdXvCOYuWbJiKLTWo >
2019-06-12 00:50:27     gentz   (I haven't tested if it compiles)
2019-06-12 00:50:43     gentz   What do the logs say, UndeadLeech
2019-06-12 00:51:02     gentz   Is there a correlation between the swap and the delay?
2019-06-12 00:51:18     gentz   Does it still happen?
2019-06-12 00:52:06     gentz   Maybe log both when the swap starts and ends.
2019-06-12 00:58:28     UndeadLeech     Tough to say since both happen so often.
2019-06-12 00:59:30     gentz   Can you upload the log (with both swap ends and starts logged)?
2019-06-12 01:00:17     UndeadLeech     One second, let me go into a bit more detail first.
2019-06-12 01:02:16     UndeadLeech     https://paste.rs/6sG
2019-06-12 01:02:21     UndeadLeech     I think this should be sufficient^
2019-06-12 01:02:47     UndeadLeech     However, I just noticed that it always receives characters when the swap is in progress though, never when it's done.
2019-06-12 01:03:40     gentz   GLX in use?
2019-06-12 01:04:19     UndeadLeech     If I move the `context.swap_buffers()` below the `tx.send(()).unwrap()` I get the same results though, even though it's start -> end then instead of the other way around.
2019-06-12 01:04:28     UndeadLeech     How do I tell if it's GLX?
2019-06-12 01:05:17     gentz   `get_egl_display` is None.
2019-06-12 01:05:26     UndeadLeech     On the context?
2019-06-12 01:05:47     gentz   Yeah
2019-06-12 01:05:48     gentz   https://docs.rs/glutin/0.21.0/glutin/os/trait.ContextTraitExt.html#tymethod.get_egl_display
2019-06-12 01:06:22     UndeadLeech     Yeah, it's None.
2019-06-12 01:08:08     gentz   `cat  /tmp/Log{1,2} | sort`? Can I see a small section of that, timestamps included?
2019-06-12 01:10:55     UndeadLeech     I can give you all of it: https://paste.rs/Q13
2019-06-12 01:11:23     UndeadLeech     Want me to do it with swap start/end?
2019-06-12 01:11:47     gentz   Nah, just give me a while to write up an xlib example
2019-06-12 01:12:09     gentz   Maybe annotate winit's eventloop function with a similar logger and see where the most time is being spent?
2019-06-12 01:13:14     UndeadLeech     Yeah, looking into winit would be a reasonable next step, though it would likely take me a while to get done.
2019-06-12 01:13:46     UndeadLeech     I'm still not sure if there isn't some fundamental logical error with things getting locked up.
2019-06-12 01:16:15     gentz   You said it only breaks after the swap finishes.
2019-06-12 01:16:18     gentz   Sounds like an xlib issue to me
2019-06-12 01:16:53     gentz   If it is a logic error in glutin/winit, we'd expect no events to arrive between the swap start and end
2019-06-12 01:19:04     UndeadLeech     I feel like I should be able to repro with winit/glutin examples then though.
```

</details>

However I'm not entirely sure if this problem is related, since the maximum amount of delay would be a single frame. In general it seems like the delay of events while swapping buffers might be a bit of a red herring, since in theory it shouldn't make any difference if events come in all bundled right before the next buffer swap or if they come in during the last swap, as long as the event loop is completely drained before the next swap.

**Comment by ZeGentzy** [comment]
Most likely fixed by #1048. Closing.
